utilities/include/vec3.hpp

```diff
@@ -1,120 +1,131 @@
+#ifndef VEC3.HPP
+#define VEC3.HPP
+
 //code by Ameya Vikrama Singh
 #include<iostream>
 #include<cmath>
 class vec3{
+    float x,y,z;
     public:
-        float x,y,z;
-
-        vec3(){
-            x=y=z=0;
-        }
+        vec3() : x{0},y{0},z{0}{}
 
         //Cartesian coordinate Constructor
-        vec3(float x,float y,float z){
-            this->x=x;
-            this->y=y;
-            this->z=z;
-        }
-        
-        //assignment operator
-        void operator=(vec3 s){
-            x=s.x;
-            y=s.y;
-            z=s.z;
-        }
+        vec3(float X,float Y,float Z) : x{X}, y{Y}, z{Z}{}
 
         //norm function
-        float norm(){
+        constexpr float norm() const{
             return sqrt(x*x+y*y+z*z);
         }
 
         //vector addition
-        vec3 operator+(vec3 s){
-            vec3 sum;
-            sum.x=s.x+x;
-            sum.y=s.y+y;
-            sum.z=s.z+z;
-            return sum;
-        }
+        friend vec3 operator+(vec3 const& h, vec3 const & s);
 
         //vector subtraction
-        vec3 operator-(vec3 s){
-            vec3 diff;
-            diff.x=x-s.x;
-            diff.y=y-s.y;
-            diff.z=z-s.z;
-            return diff;
-        }
+        friend vec3 operator-(vec3 const& h, vec3 const & s);
+
+        //scalar post-multiplication
+        friend vec3 operator*(vec3 const& h, float const & s);
+
+        //scalar post-division
+        friend vec3 operator/(vec3 const& h, float const & s);
+
 
         //addition+assignemnt
-        void operator+=(vec3 s){
+        vec3& operator+=(vec3 const & s){
```

Comment:
For the sake of homogeneity, you should have all the operator overloads as friend functions, or none of them. Don't have some as friends and others as member functions. Pick one and stick with it.

Reply:
If I define compound assignment as friend, it cant modify x,y,z because they are private right? So should I make all of them members?